### ISSUE TITLE
NullPointerException when running grobid-core's processXXX commands

### ISSUE THREAD

**Issue by liar666** [kermitt2/grobid#11]
When running grobid-core's commands "processXXX", I get a NullPointerException: - processDate (ProcessEngine:103)
- processAuthorsHeader (ProcessEngine:117)
- processAuthorsCitation (ProcessEngine:131)
  The line that causes the problem is:
  Utilities.writeInFile(pGbdArgs.getPath2Output() + File.separator + "result", result.get(0).toTEI());
  I'm not sure if it comes from the result being "null" because I use incorrectly the commands (in this case, maybe it would be worth explaining more the useage in the docs), or if it comes from attempting to write to a file, when the commands in the doc seems not to configure a file to write to (is it a problem in the commands given in the doc or a larger problem of attempting to write to a file when the call of the commands from the command-line should write to the terminal?).

I also get and error for processAffiliation, that comes from the line: processAffiliation (AffiliationAddressParser:28), which reads:
input.trim();
I think the error is related to the above ones, but I'm not sure for the moment. I believe 'input' might be "null" here because i) it is not captured from the commands line arguments, ii) I have not understood how to pass the argument to the command, iii) there are some pre-treatments done on the argument that nullify the string.

**Comment by lfoppiano** [comment]
@liar666 would you mind to test again this issue? I believe it shoudl be solved it. If not I'm ready to takle it quickly.

**Comment by liar666** [comment]
Hi,

I've tried it again and it seems not to crash anymore. Unfortunately, it does not seems to return anything either :[

**Comment by lfoppiano** [comment]
Could you provide me the input files that are causing the error/behaviour?

**Comment by liar666** [comment]
Unfortunately, I don't remember exactly what I did when I opened the issue and I couldn't find back the files I used :[

So, for my new tests, I used mostly extracts from:
http://www.cs.toronto.edu/~graves/preprint.pdf

$ java -Xmx1024m -jar grobid-core/target/grobid-core-0.4.1-SNAPSHOT.one-jar.jar -gH grobid-home/ -gP grobid-home/config/grobid.properties -exe processAuthorsHeader -s "Vojislav Kecman, The University of Auckland, Auckland, NZ"

$ java -Xmx1024m -jar grobid-core/target/grobid-core-0.4.1-SNAPSHOT.one-jar.jar -gH grobid-home/ -gP grobid-home/config/grobid.properties -exe processAuthorsCitation -s "Y. Bengio. A connectionist approach to speech recognition. International Journal on Pattern Recognition and Artificial Intelligence, 7(4):647–668, 1993. URL http://www.iro.umontreal.ca/~lisa/pointeurs/ijprai93.ps.\nY. Bengio. Markovian models for sequential data. Neural Computing Surveys, 2:129–162, 1999. URL http://www.iro.umontreal.ca/~lisa/pointeurs/hmms.ps."

I also tried the following command and got nothing too:
$ java -Xmx1024m -jar grobid-core/target/grobid-core-0.4.1-SNAPSHOT.one-jar.jar -gH grobid-home/ -gP grobid-home/config/grobid.properties -exe processRawReference -s "Y. Bengio. A connectionist approach to speech recognition. International Journal on Pattern Recognition and Artificial Intelligence, 7(4):647–668, 1993. URL http://www.iro.umontreal.ca/~lisa/pointeurs/ijprai93.ps.\nY. Bengio. Markovian models for sequential data. Neural Computing Surveys, 2:129–162, 1999. URL http://www.iro.umontreal.ca/~lisa/pointeurs/hmms.ps."

However, I'm not sure I'm using GROBID correctly, since it's been a long time I haven't used it and I've re-read only very quickly the docs here:
https://grobid.readthedocs.io/en/latest/Grobid-batch/#processauthorscitation

Cheers

BTW: there are multiple entries in the cited doc for which the list of "needed parameters" and "example" lines are not in coherence :)

**Comment by lfoppiano** [comment]
@liar666 Thanks for your feedback. We will have a look at it and correct the documentation ;)

**Comment by kermitt2** [comment]
@liar666 Hi, I hope you're going well! (let's try one day to meet in Paris by the way!?)

Actuallly your input strings are not consistent for the methods you are calling. 

The batch ```processAuthorsHeader``` normally runs only a sequence of author names as input, so from your example: 
```bash
$ java -Xmx1024m -jar grobid-core/target/grobid-core-0.4.1-SNAPSHOT.one-jar.jar -gH grobid-home/ -gP grobid-home/config/grobid.properties -exe processAuthorsHeader -s "Vojislav Kecman"
```
is fine and will return the structure author name. 

```
<persName><forename type="first">Vojislav</forename><surname>Kecman</surname></persName>
```

However, the batch processAuthorsHeader does not process complete author+affiliation sequence, neither complete citation string. For processing a complete citation string, ```processRawReference``` should be used: 

```bash
$ java -Xmx1024m -jar grobid-core/target/grobid-core-0.4.1-SNAPSHOT.one-jar.jar -gH grobid-home/ -gP grobid-home/config/grobid.properties -exe processRawReference -s "Y. Bengio. A connectionist approach to speech recognition. International Journal on Pattern Recognition and Artificial Intelligence, 7(4):647–668, 1993. URL http://www.iro.umontreal.ca/~lisa/pointeurs/ijprai93.ps.\nY."
```

This is working fine, and gives the author structure, together with the citation fields. 

```
<biblStruct  >
        <analytic>
                <title level="a" type="main">A connectionist approach to speech recognition</title>
                <author>
                        <persName>
                                <forename type="first">Y</forename>
                                <surname>Bengio</surname>
                        </persName>
                </author>
        </analytic>
        <monogr>
                <title level="j">International Journal on Pattern Recognition and Artificial Intelligence</title>
                <imprint>
                        <biblScope unit="volume">7</biblScope>
                        <biblScope unit="issue">4</biblScope>
                        <biblScope unit="page" from="647" to="668" />
                        <date type="published" when="1993" />
                </imprint>
        </monogr>
</biblStruct>
```

The ```processRawReference``` exe only processes one reference string at the time. Otherwise you would need to call first the method for segmenting several bibliographical references... which is currenlty not possible to call with the command line on simply a string (see issue  #71) - it works now from PDF only. So your examples with several references at the same time in the input strings are not appropriate for the batch method ```processRawReference```.

So to summarize, currently when the input is not consistent with the processing method that is called, the batch process will return either nothing, or incorrect structures (unfortunately rather the second one). We could maybe try to clarify the constraints on the input strings in the documentation. 

As for the original issue related to null point exception, I think I have put more checks on empty and invalid inputs in the last years, but I am not sure to have covered all cases. 

About the multiple entries in the cited doc for which the list of "needed parameters" and "example" lines are not in consistent, could you maybe you point more specifically to cases ? :D

**Comment by liar666** [comment]
On 09/05/2016 23:45, Patrice Lopez wrote:
> @liar666 Hi, I hope you're going well! (let's try one day to meet in Paris by the way!?)

Hey thanks, I hope you're dong well too. FYI, my company has moved to the center of Paris, but if you are in the neighborhood, one day, let's have a beer :)

> Actually your input strings are not consistent for the methods you are calling. 

OK Thanks for the info, that's what I was thinking ;) I'll try to put these more complete examples in GROBID's online doc this afternoon. 

-- 
Guillaume MULLER, PhD
Presans c/o REMIX COWORKING - L'APPART
57 rue de Turbigo
75003 Paris
France
http://www.presans.com
http://feeds.feedburner.com/OYI/fr
